### PR TITLE
Adding software filter rule

### PR DESCRIPTION
**Pull request by YuriiPerets**

### CODE REVIEW

File: src/bashi/filter_software_dependency.py [hunk 1]
```diff
@@ -40,4 +42,11 @@ def software_dependency_filter(
     Returns:
         bool: True, if parameter-value-tuple is valid.
     """
+
+    for compiler in (HOST_COMPILER, DEVICE_COMPILER):
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:46]: A comment is missing, what the rule is doing. Also a rule number. Should be `d1`.

File: src/bashi/filter_software_dependency.py [hunk 2]
```diff
@@ -40,4 +42,11 @@ def software_dependency_filter(
...
+        if compiler in row and row[compiler].name == GCC:
+            if row[compiler].version == pkv.parse("6"):
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:48]:
```suggestion
            if row[compiler].version <= pkv.parse("6"):
```

**Review by SimeonEhrig — APPROVED** [review]

**Comment by SimeonEhrig** [comment]
Before-after-tests with small, handwritten test sets are missing. See an example here: https://github.com/alpaka-group/bashi/pull/17/files

**Comment by SimeonEhrig** [comment]
Starting point for the filter rule test, create the file `test_filter_software_dependency.py` in the folder `tests` with the following content and extend it:

```python
# pylint: disable=missing-docstring
import unittest
import io

from collections import OrderedDict as OD
from utils_test import parse_param_val as ppv
from bashi.globals import *  # pylint: disable=wildcard-import,unused-wildcard-import
from bashi.filter_software_dependency import software_dependency_filter_typechecked


class TestOldGCCVersionInUbuntu2004(unittest.TestCase):
    def test_valid_gcc_is_in_ubuntu_2004_d1(self):
        for gcc_version in [7, 13, 99]:
            self.assertTrue(
                software_dependency_filter_typechecked(
                    OD({HOST_COMPILER: ppv((GCC, gcc_version)), UBUNTU: ppv((UBUNTU, "20.04"))}),
                )
            )

            self.assertTrue(
                software_dependency_filter_typechecked(
                    OD(
                        {
                            HOST_COMPILER: ppv((GCC, gcc_version)),
                            DEVICE_COMPILER: ppv((GCC, gcc_version)),
                            UBUNTU: ppv((UBUNTU, "20.04")),
                        }
                    ),
                )
            )

    def test_not_valid_gcc_is_in_ubuntu_2004_d1(self):
        for gcc_version in [6, 3, 1]:
            reason_msg = io.StringIO()
            self.assertFalse(
                software_dependency_filter_typechecked(
                    OD({HOST_COMPILER: ppv((GCC, gcc_version)), UBUNTU: ppv((UBUNTU, "20.04"))}),
                    reason_msg,
                ),
                f"GCC {gcc_version} + Ubuntu 20.04",
            )
            self.assertEqual(
                reason_msg.getvalue(),
                f"GCC {gcc_version} is not available in Ubuntu 20.04",
            )
```

**Comment by SimeonEhrig** [comment]
Add the following test to the `test_results.py` and extend it.

```python
    def test_remove_unsupported_gcc_versions_for_ubuntu2004(self):
        test_param_value_pairs: List[ParameterValuePair] = parse_expected_val_pairs(
            [
                OD({HOST_COMPILER: (GCC, 10), DEVICE_COMPILER: (NVCC, 11.2)}),
                OD({HOST_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "20.04")}),
                OD({HOST_COMPILER: (GCC, 5), UBUNTU: (UBUNTU, "20.04")}),
                OD({HOST_COMPILER: (GCC, 7), UBUNTU: (UBUNTU, "20.04")}),
                OD({DEVICE_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "20.04")}),
                OD({HOST_COMPILER: (GCC, 6), DEVICE_COMPILER: (GCC, 6)}),
            ]
        )

        _remove_unsupported_gcc_versions_for_ubuntu2004(test_param_value_pairs)

        test_param_value_pairs.sort()
        expected_results = sorted(
            parse_expected_val_pairs(
                [
                    OD({HOST_COMPILER: (GCC, 10), DEVICE_COMPILER: (NVCC, 11.2)}),
                    OD({HOST_COMPILER: (GCC, 7), UBUNTU: (UBUNTU, "20.04")}),
                    OD({HOST_COMPILER: (GCC, 6), DEVICE_COMPILER: (GCC, 6)}),
                ]
            )
        )

        self.assertEqual(
            test_param_value_pairs,
            expected_results,
            create_diff_parameter_value_pairs(test_param_value_pairs, expected_results),
        )
```

**Comment by SimeonEhrig** [comment]
For the test input, there are some general rules, what you should test:

- corner cases like gcc 6 (first unsupported version) and gcc 7 (first supported version)
- versions in both directions of the corner case, like gcc 1 and 3 (old versions) and 9, 13 (released version) and 99 (potential upcoming version, which could be released in future)
- different number of input parameter
  - minimum case like host compiler + Ubuntu, device compiler + Ubuntu and host compiler + device compiler
  - all relevant parameter together: host and device compiler and Ubuntu
  - test with unrelated parameter e.g.   host compiler + Ubuntu + CMake
  - - test with fully unrelated parameter e.g.  Boost + CMake

File: tests/test_results.py [hunk 3]
```diff
@@ -2202,3 +2203,50 @@ def test_remove_unsupported_clang_sdk_versions_for_clang_cuda(self):
...
+                    OD({HOST_COMPILER: (GCC, 3), DEVICE_COMPILER: (NVCC, 12.2)}),
+                ]
+            )
+        )
+
+        self.assertEqual(
+            test_param_value_pairs,
+            expected_results,
+            create_diff_parameter_value_pairs(test_param_value_pairs, expected_results),
+        )
```
Comment (SimeonEhrig) [inline tests/test_results.py:2252]:
```suggestion
    def test_remove_unsupported_gcc_versions_for_ubuntu2004_and_later(self):
        test_param_value_pairs: List[ParameterValuePair] = parse_expected_val_pairs(
            [
                OD({DEVICE_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "20.04")}),
                OD({HOST_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "20.04")}),
                OD({DEVICE_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "22.04")}),
                OD({HOST_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "22.04")}),
                OD({HOST_COMPILER: (GCC, 3), UBUNTU: (UBUNTU, "22.04")}),
                OD({HOST_COMPILER: (GCC, 12), UBUNTU: (UBUNTU, "22.04")}),
                OD({HOST_COMPILER: (GCC, 7), UBUNTU: (UBUNTU, "20.04")}),
                OD({HOST_COMPILER: (GCC, 3), UBUNTU: (UBUNTU, "20.04")}),
                OD({HOST_COMPILER: (GCC, 13), UBUNTU: (UBUNTU, "20.04")}),
                OD({HOST_COMPILER: (GCC, 99), UBUNTU: (UBUNTU, "20.04")}),
                OD({DEVICE_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "18.04")}),
                OD({DEVICE_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "18.04")}),
                OD({HOST_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "18.04")}),
                OD({HOST_COMPILER: (GCC, 7), UBUNTU: (UBUNTU, "18.04")}),
                OD({HOST_COMPILER: (GCC, 3), UBUNTU: (UBUNTU, "18.04")}),
                OD({HOST_COMPILER: (GCC, 6), DEVICE_COMPILER: (GCC, 6)}),
                OD({HOST_COMPILER: (GCC, 10), DEVICE_COMPILER: (NVCC, 11.2)}),
                OD({HOST_COMPILER: (GCC, 6), DEVICE_COMPILER: (NVCC, 11.2)}),
                OD({HOST_COMPILER: (GCC, 3), DEVICE_COMPILER: (NVCC, 12.2)}),
            ]
        )
        expected_results = parse_expected_val_pairs(
            [
                OD({HOST_COMPILER: (GCC, 7), UBUNTU: (UBUNTU, "20.04")}),
                OD({HOST_COMPILER: (GCC, 12), UBUNTU: (UBUNTU, "22.04")}),
                OD({HOST_COMPILER: (GCC, 13), UBUNTU: (UBUNTU, "20.04")}),
                OD({HOST_COMPILER: (GCC, 99), UBUNTU: (UBUNTU, "20.04")}),
                OD({DEVICE_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "18.04")}),
                OD({HOST_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "18.04")}),
                OD({DEVICE_COMPILER: (GCC, 6), UBUNTU: (UBUNTU, "18.04")}),
                OD({HOST_COMPILER: (GCC, 7), UBUNTU: (UBUNTU, "18.04")}),
                OD({HOST_COMPILER: (GCC, 3), UBUNTU: (UBUNTU, "18.04")}),
                OD({HOST_COMPILER: (GCC, 6), DEVICE_COMPILER: (GCC, 6)}),
                OD({HOST_COMPILER: (GCC, 10), DEVICE_COMPILER: (NVCC, 11.2)}),
                OD({HOST_COMPILER: (GCC, 6), DEVICE_COMPILER: (NVCC, 11.2)}),
                OD({HOST_COMPILER: (GCC, 3), DEVICE_COMPILER: (NVCC, 12.2)}),
            ]
        )
        default_remove_test(
            _remove_unsupported_gcc_versions_for_ubuntu2004,
            test_param_value_pairs,
            expected_results,
            self,
        )
```

File: src/bashi/results.py [hunk 4]
```diff
@@ -641,3 +641,23 @@ def filter_func(param_val_pair: ParameterValuePair) -> bool:
...
+    for compiler_type in (HOST_COMPILER, DEVICE_COMPILER):
+        remove_parameter_value_pairs(
+            parameter_value_pairs,
+            parameter1=compiler_type,
+            value_name1=GCC,
+            value_version1=">6",
+            parameter2=UBUNTU,
+            value_name2=UBUNTU,
+            value_version2="==20.04",
+        )
```
Comment (SimeonEhrig) [inline src/bashi/results.py:663]:
```suggestion
for compiler_type in (HOST_COMPILER, DEVICE_COMPILER):
        for gcc_version in range(1, 7):
            remove_parameter_value_pairs(
                parameter_value_pairs,
                removed_parameter_value_pairs,
                parameter1=compiler_type,
                value_name1=GCC,
                value_version1=gcc_version,
                parameter2=UBUNTU,
                value_name2=UBUNTU,
                value_version2="<20.04",
            )
```

Is not doable with two version ranges. Fortunately, released GCC versions are countable. Therefore we can remove each of the version by hand.

File: tests/test_filter_software_dependency.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import io
+
+from collections import OrderedDict as OD
+from utils_test import parse_param_val as ppv
+from bashi.globals import *  # pylint: disable=wildcard-import,unused-wildcard-import
+from bashi.filter_software_dependency import software_dependency_filter_typechecked
+
+
+class TestOldGCCVersionInUbuntu2004(unittest.TestCase):
+    def test_valid_gcc_is_in_ubuntu_2004_d1(self):
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:12]: Please add somewhere in your test cases also tests with Ubuntu 18.04, 22.04 and 24.04

File: src/bashi/results.py [hunk 6]
```diff
@@ -747,3 +749,26 @@ def filter_func(param_val_pair: ParameterValuePair) -> bool:
...
+        for gcc_version in range(1, 7):
+            remove_parameter_value_pairs(
+                parameter_value_pairs,
+                removed_parameter_value_pairs,
+                parameter1=compiler_type,
+                value_name1=GCC,
+                value_version1=gcc_version,
+                parameter2=UBUNTU,
+                value_name2=UBUNTU,
+                value_version2="<=20.04",
```
Comment (SimeonEhrig) [inline src/bashi/results.py:773]:
```suggestion
                value_version2="<20.04",
```

File: src/bashi/filter_software_dependency.py [hunk 7]
```diff
@@ -40,4 +43,85 @@ def software_dependency_filter(
...
+        if DEVICE_COMPILER in row and row[DEVICE_COMPILER].name == GCC:
+            if HOST_COMPILER in row and row[HOST_COMPILER].name == GCC:
+                for gcc_version in range(1, 7):
+                    if (
+                        row[HOST_COMPILER].version == gcc_version
+                        or row[DEVICE_COMPILER].version == gcc_version
+                    ):
+                        reason(
+                            output,
+                            "GCC as host and device compiler older than 6 version is not supported on Ubuntu 20.04 or newer",
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:123]:
```suggestion
                            f"GCC {row[DEVICE_COMPILER].version} is not available in Ubuntu {__ubuntu_version_to_string(row[UBUNTU].version)}",",
```

`__ubuntu_version_to_string` is defined as follow:
```python
def __ubuntu_version_to_string(version: pkv.Version) -> str:
    return f"{version.major}.{version.minor:02}"
```

File: tests/test_filter_software_dependency.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+                    OD({HOST_COMPILER: ppv((GCC, gcc_version)), UBUNTU: ppv((UBUNTU, "20.04"))}),
+                    reason_msg,
+                ),
+                f"GCC {gcc_version} + Ubuntu 20.04",
+            )
+            self.assertEqual(
+                reason_msg.getvalue(),
+                f"GCC {gcc_version} is not available in Ubuntu 20.04",
+            )
+            self.assertFalse(
```
Comment (SimeonEhrig) [inline tests/test_filter_software_dependency.py:81]: You need to clean the `reason_msg` before you use again. Otherwise you will have two error messages in the buffer this call.

You can ether write `reason_msg = io.StringIO()` to create a new buffer or use two different buffers with the names `reason_msg1` and `reason_msg2`. I suggest the second solution to be sure, that the buffer is empty.

File: src/bashi/filter_software_dependency.py [hunk 9]
```diff
@@ -40,4 +43,85 @@ def software_dependency_filter(
...
+                    reason(
+                        output,
+                        "GCC as host and device compiler older than 6 version is not supported on Ubuntu 20.04 or newer",
+                    )
+                    
+                    
+
+"""
+    if UBUNTU in row and row[UBUNTU].version >= pkv.parse("20.04"):
+        if DEVICE_COMPILER in row and row[DEVICE_COMPILER].name == GCC:
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:114]: This hurts the rule of 2: https://github.com/alpaka-group/bashi/blob/main/docs/rules.md#write-only-rules-with-two-parameters
Before you filter, you check if the three parameter Ubuntu, host compiler and device compiler are already set. But this hurt the rule of 2 and in reality you can already cancel the combination, if you know the ubuntu version and if the host **or** device compiler is GCC 6 or older. 

Therefore you need to check with:

```python
if UBUNTU in row and row[UBUNTU].version >= pkv.parse("20.04"):
        if DEVICE_COMPILER in row and row[DEVICE_COMPILER].name == GCC:
             # ...
        if HOST_COMPILER in row and row[HOST_COMPILER].name == GCC:
            # ...
```

File: src/bashi/filter_software_dependency.py [hunk 10]
```diff
@@ -40,4 +43,85 @@ def software_dependency_filter(
...
+            if HOST_COMPILER in row and row[HOST_COMPILER].name == GCC:
+                for gcc_version in range(1, 7):
+                    if (
+                        row[HOST_COMPILER].version == gcc_version
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:118]: Unfortunately, the weak type system of python is not your friend here. You compare a `packaging.version.Version` with a `int`, which is always `False` because they have different types.

You need to parse the version with `pkv.parse(gcc_version)`. Instead checking for specific version, you can check for a range.

 ```suggestion
                         row[HOST_COMPILER].version <= pkv.parse("6")
```

File: src/bashi/filter_software_dependency.py [hunk 11]
```diff
@@ -40,4 +43,23 @@ def software_dependency_filter(
...
+                )
+                return False
+        if HOST_COMPILER in row and row[HOST_COMPILER].name == GCC:
+            if row[HOST_COMPILER].version <= pkv.parse("6"):
+                reason(
+                    output,
+                    f"Host compiler GCC {row[HOST_COMPILER].version}"
+                    f" is not available in Ubuntu {row[UBUNTU].version}",
+                )
+                return False
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:63]:
```suggestion
    # Rule: d1
    # GCC 6 and older is not available in Ubuntu 20.04 and newer
    if UBUNTU in row and row[UBUNTU].version >= pkv.parse("20.04"):
        for compiler_type in (HOST_COMPILER, DEVICE_COMPILER):
            if compiler_type in row and row[compiler_type].name == GCC:
                if row[compiler_type].version <= pkv.parse("6"):
                    reason(
                        output,
                        f"{__pretty_name_compiler(compiler_type)} GCC {row[compiler_type].version} "
                        f"is not available in Ubuntu {__ubuntu_version_to_string(row[UBUNTU].version)}",
                    )
                    return False
```

File: src/bashi/filter_software_dependency.py [hunk 12]
```diff
@@ -8,8 +8,11 @@
 """
 
 from typing import Optional, IO
+import packaging.version as pkv
 from typeguard import typechecked
 from bashi.types import ParameterValueTuple
+from bashi.globals import DEVICE_COMPILER, HOST_COMPILER, GCC, UBUNTU
+from bashi.utils import reason
```
Comment (SimeonEhrig) [inline src/bashi/filter_software_dependency.py:17]:
```suggestion
def __ubuntu_version_to_string(version: pkv.Version) -> str:
    """Returns the Ubuntu version representation correctly. Ubuntu versions
    use a leading 0 in their version scheme for months before October. pkv.parse()`
    parses e.g. the 04 from 20.04 to 4. Therefore the string representation of
    str(pkv.parse(“20.04”)) is `20.4`. This function returns the correct version scheme.
    For Ubuntu `20.04` it is `20.04`.

    Args:
        version (pkv.Version): Ubuntu version

    Returns:
        str: string representation of the Ubuntu version
    """
    return f"{version.major}.{version.minor:02}"


def __pretty_name_compiler(constant: str) -> str:
    """Returns the string representation of the constants HOST_COMPILER and DEVICE_COMPILER in a
    human-readable version.

    Args:
        constant (str): Ether HOST_COMPILER or DEVICE_COMPILER

    Returns:
        str: human-readable string representation of HOST_COMPILER or DEVICE_COMPILER
    """
    if constant == HOST_COMPILER:
        return "host compiler"
    if constant == DEVICE_COMPILER:
        return "device compiler"
    return "unknown compiler type"

```

**Review by SimeonEhrig — CHANGES_REQUESTED** [review]

**Comment by SimeonEhrig** [comment]
Please remove the comments on top of the function `software_dependency_filter()`. They are not required anymore, because the filter function has functionality now. 

```diff
- # TODO(SimeonEhrig): remove disable=unused-argument
- # only required for the CI at the moment
def software_dependency_filter(
```

**Review by SimeonEhrig — APPROVED** [review]
